File: internal/modules/courses/content.go
package courses

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"net/url"
	"strings"
	"unicode/utf8"
)

const (
	LessonContentSchemaVersion = 1
	MaxLessonContentBytes      = 1 << 20
	MaxLessonBlocks            = 200
	MaxBlockTextCharacters     = 50000
	MaxCodeCharacters          = 100000
	MaxURLLength               = 2048
	MaxTableColumns            = 20
	MaxTableRows               = 200
)

type BlockType string

const (
	BlockText           BlockType = "TEXT"
	BlockHeading        BlockType = "HEADING"
	BlockImage          BlockType = "IMAGE"
	BlockVideo          BlockType = "VIDEO"
	BlockAudio          BlockType = "AUDIO"
	BlockCode           BlockType = "CODE"
	BlockQuote          BlockType = "QUOTE"
	BlockCallout        BlockType = "CALLOUT"
	BlockTable          BlockType = "TABLE"
	BlockDownload       BlockType = "DOWNLOAD"
	BlockKnowledgeCheck BlockType = "KNOWLEDGE_CHECK"
	BlockDivider        BlockType = "DIVIDER"
)

type LessonContent struct {
	SchemaVersion int     `json:"schemaVersion"`
	Blocks        []Block `json:"blocks"`
}

func (c LessonContent) Validate() error {
	if c.SchemaVersion != 1 {
		return errors.New("unsupported lesson content schema version")
	}
	if len(c.Blocks) > MaxLessonBlocks {
		return errors.New("too many lesson blocks")
	}
	seen := map[string]bool{}
	for _, b := range c.Blocks {
		k, e := NormalizeStructureKey(b.Key)
		if e != nil || k != b.Key {
			return errors.New("invalid lesson block key")
		}
		if seen[b.Key] {
			return errors.New("duplicate lesson block key")
		}
		seen[b.Key] = true
		if e := b.Validate(); e != nil {
			return e
		}
	}
	return nil
}
func MarshalLessonContent(c LessonContent) ([]byte, error) {
	if e := c.Validate(); e != nil {
		return nil, e
	}
	b, e := json.Marshal(c)
	if e != nil || len(b) > MaxLessonContentBytes {
		return nil, errors.New("invalid lesson content")
	}
	return b, nil
}
func ParseLessonContent(b []byte) (LessonContent, error) {
	if len(b) == 0 || len(b) > MaxLessonContentBytes {
		return LessonContent{}, errors.New("invalid lesson content size")
	}
	var c LessonContent
	if e := strict(b, &c); e != nil {
		return LessonContent{}, errors.New("invalid lesson content document")
	}
	return c, c.Validate()
}

type BlockPayload interface {
	blockPayload()
	Validate() error
}
type Block struct {
	Key     string
	Type    BlockType
	Payload BlockPayload
}

func (b Block) Validate() error {
	switch b.Type {
	case BlockText:
		return expect[TextBlockPayload](b.Payload)
	case BlockHeading:
		return expect[HeadingBlockPayload](b.Payload)
	case BlockImage:
		return expect[ImageBlockPayload](b.Payload)
	case BlockVideo:
		return expect[VideoBlockPayload](b.Payload)
	case BlockAudio:
		return expect[AudioBlockPayload](b.Payload)
	case BlockCode:
		return expect[CodeBlockPayload](b.Payload)
	case BlockQuote:
		return expect[QuoteBlockPayload](b.Payload)
	case BlockCallout:
		return expect[CalloutBlockPayload](b.Payload)
	case BlockTable:
		return expect[TableBlockPayload](b.Payload)
	case BlockDownload:
		return expect[DownloadBlockPayload](b.Payload)
	case BlockKnowledgeCheck:
		return expect[KnowledgeCheckBlockPayload](b.Payload)
	case BlockDivider:
		return expect[DividerBlockPayload](b.Payload)
	}
	return errors.New("unsupported lesson block type")
}
func expect[T BlockPayload](p BlockPayload) error {
	v, ok := p.(T)
	if !ok {
		return errors.New("block payload does not match type")
	}
	return v.Validate()
}
func (b Block) MarshalJSON() ([]byte, error) {
	if e := b.Validate(); e != nil {
		return nil, e
	}
	return json.Marshal(struct {
		Key     string       `json:"key"`
		Type    BlockType    `json:"type"`
		Payload BlockPayload `json:"payload"`
	}{b.Key, b.Type, b.Payload})
}
func (b *Block) UnmarshalJSON(data []byte) error {
	var r struct {
		Key     string          `json:"key"`
		Type    BlockType       `json:"type"`
		Payload json.RawMessage `json:"payload"`
	}
	if e := strict(data, &r); e != nil || len(r.Payload) == 0 {
		return errors.New("invalid lesson block")
	}
	p, e := blockPayload(r.Type, r.Payload)
	if e != nil {
		return e
	}
	b.Key, b.Type, b.Payload = r.Key, r.Type, p
	return nil
}
func blockPayload(t BlockType, b []byte) (BlockPayload, error) {
	switch t {
	case BlockText:
		return decode[TextBlockPayload](b)
	case BlockHeading:
		return decode[HeadingBlockPayload](b)
	case BlockImage:
		return decode[ImageBlockPayload](b)
	case BlockVideo:
		return decode[VideoBlockPayload](b)
	case BlockAudio:
		return decode[AudioBlockPayload](b)
	case BlockCode:
		return decode[CodeBlockPayload](b)
	case BlockQuote:
		return decode[QuoteBlockPayload](b)
	case BlockCallout:
		return decode[CalloutBlockPayload](b)
	case BlockTable:
		return decode[TableBlockPayload](b)
	case BlockDownload:
		return decode[DownloadBlockPayload](b)
	case BlockKnowledgeCheck:
		return decode[KnowledgeCheckBlockPayload](b)
	case BlockDivider:
		return decode[DividerBlockPayload](b)
	}
	return nil, errors.New("unsupported lesson block type")
}
func decode[T BlockPayload](b []byte) (T, error) { var v T; return v, strict(b, &v) }
func strict(b []byte, v any) error {
	d := json.NewDecoder(bytes.NewReader(b))
	d.DisallowUnknownFields()
	if e := d.Decode(v); e != nil {
		return e
	}
	var extra any
	if e := d.Decode(&extra); e != io.EOF {
		return errors.New("trailing JSON")
	}
	return nil
}

type RichText struct {
	Nodes []RichTextNode `json:"nodes"`
}
type RichTextNode struct {
	Type    string             `json:"type"`
	Content []RichTextInline   `json:"content,omitempty"`
	Items   [][]RichTextInline `json:"items,omitempty"`
}
type RichTextInline struct {
	Type  string         `json:"type"`
	Text  string         `json:"text,omitempty"`
	Marks []RichTextMark `json:"marks,omitempty"`
}
type RichTextMark struct {
	Type string `json:"type"`
	Href string `json:"href,omitempty"`
}

func (r RichText) Validate() error {
	if len(r.Nodes) == 0 || len(r.Nodes) > 100 {
		return errors.New("invalid rich text")
	}
	n := 0
	for _, x := range r.Nodes {
		switch x.Type {
		case "paragraph":
			c, e := inlines(x.Content)
			if e != nil || len(x.Items) > 0 {
				return errors.New("invalid paragraph")
			}
			n += c
		case "bullet_list", "ordered_list":
			if len(x.Content) > 0 || len(x.Items) == 0 || len(x.Items) > 100 {
				return errors.New("invalid list")
			}
			for _, i := range x.Items {
				c, e := inlines(i)
				if e != nil {
					return e
				}
				n += c
			}
		default:
			return errors.New("unsupported rich text node")
		}
		if n > MaxBlockTextCharacters {
			return errors.New("rich text too large")
		}
	}
	if n == 0 {
		return errors.New("rich text requires text")
	}
	return nil
}
func inlines(x []RichTextInline) (int, error) {
	if len(x) == 0 || len(x) > 200 {
		return 0, errors.New("invalid rich text inline")
	}
	n := 0
	for _, i := range x {
		if i.Type == "hard_break" {
			if i.Text != "" || len(i.Marks) > 0 {
				return 0, errors.New("invalid hard break")
			}
			continue
		}
		if i.Type != "text" || i.Text == "" || len(i.Marks) > 4 {
			return 0, errors.New("invalid rich text inline")
		}
		n += utf8.RuneCountInString(i.Text)
		for _, m := range i.Marks {
			if m.Type == "link" {
				if e := safeURL(m.Href); e != nil {
					return 0, e
				}
			} else if (m.Type != "emphasis" && m.Type != "strong" && m.Type != "inline_code") || m.Href != "" {
				return 0, errors.New("invalid rich text mark")
			}
		}
	}
	return n, nil
}

type TextBlockPayload struct {
	Content RichText `json:"content"`
}

func (TextBlockPayload) blockPayload()     {}
func (p TextBlockPayload) Validate() error { return p.Content.Validate() }

type HeadingBlockPayload struct {
	Level   int              `json:"level"`
	Content []RichTextInline `json:"content"`
}

func (HeadingBlockPayload) blockPayload() {}
func (p HeadingBlockPayload) Validate() error {
	if p.Level < 2 || p.Level > 4 {
		return errors.New("invalid heading level")
	}
	_, e := inlines(p.Content)
	return e
}

type AssetReference struct {
	AssetKey string `json:"assetKey"`
}

func (a AssetReference) Validate() error {
	k, e := NormalizeStructureKey(a.AssetKey)
	if e != nil || k != a.AssetKey {
		return errors.New("invalid asset reference")
	}
	return nil
}

type ImageBlockPayload struct {
	Asset      AssetReference `json:"asset"`
	AltText    string         `json:"altText,omitempty"`
	Decorative bool           `json:"decorative"`
	Caption    string         `json:"caption,omitempty"`
}

func (ImageBlockPayload) blockPayload() {}
func (p ImageBlockPayload) Validate() error {
	if e := p.Asset.Validate(); e != nil {
		return e
	}
	if len(p.AltText) > 1000 || len(p.Caption) > 4000 || (p.Decorative && p.AltText != "") || (!p.Decorative && strings.TrimSpace(p.AltText) == "") {
		return errors.New("invalid image accessibility metadata")
	}
	return nil
}

type VideoBlockPayload struct {
	Asset           AssetReference  `json:"asset"`
	Title           string          `json:"title"`
	Transcript      string          `json:"transcript,omitempty"`
	TranscriptAsset *AssetReference `json:"transcriptAsset,omitempty"`
	CaptionsAsset   AssetReference  `json:"captionsAsset"`
}

func (VideoBlockPayload) blockPayload() {}
func (p VideoBlockPayload) Validate() error {
	if e := p.Asset.Validate(); e != nil {
		return e
	}
	if e := p.CaptionsAsset.Validate(); e != nil {
		return e
	}
	if !title(p.Title) || len(p.Transcript) > MaxBlockTextCharacters || (strings.TrimSpace(p.Transcript) == "" && p.TranscriptAsset == nil) {
		return errors.New("invalid video accessibility metadata")
	}
	if p.TranscriptAsset != nil {
		return p.TranscriptAsset.Validate()
	}
	return nil
}

type AudioBlockPayload struct {
	Asset           AssetReference  `json:"asset"`
	Title           string          `json:"title"`
	Transcript      string          `json:"transcript,omitempty"`
	TranscriptAsset *AssetReference `json:"transcriptAsset,omitempty"`
}

func (AudioBlockPayload) blockPayload() {}
func (p AudioBlockPayload) Validate() error {
	if e := p.Asset.Validate(); e != nil {
		return e
	}
	if !title(p.Title) || len(p.Transcript) > MaxBlockTextCharacters || (strings.TrimSpace(p.Transcript) == "" && p.TranscriptAsset == nil) {
		return errors.New("invalid audio accessibility metadata")
	}
	if p.TranscriptAsset != nil {
		return p.TranscriptAsset.Validate()
	}
	return nil
}

type CodeBlockPayload struct {
	Code     string `json:"code"`
	Language string `json:"language,omitempty"`
	Title    string `json:"title,omitempty"`
}

func (CodeBlockPayload) blockPayload() {}
func (p CodeBlockPayload) Validate() error {
	if p.Code == "" || len(p.Code) > MaxCodeCharacters || len(p.Language) > 64 || len(p.Title) > 240 {
		return errors.New("invalid code block")
	}
	return nil
}

type QuoteBlockPayload struct {
	Text        string `json:"text"`
	Attribution string `json:"attribution,omitempty"`
	SourceURL   string `json:"sourceUrl,omitempty"`
}

func (QuoteBlockPayload) blockPayload() {}
func (p QuoteBlockPayload) Validate() error {
	if strings.TrimSpace(p.Text) == "" || len(p.Text) > MaxBlockTextCharacters || len(p.Attribution) > 240 {
		return errors.New("invalid quote")
	}
	if p.SourceURL != "" {
		return safeURL(p.SourceURL)
	}
	return nil
}

type CalloutBlockPayload struct {
	Kind    string   `json:"kind"`
	Title   string   `json:"title,omitempty"`
	Content RichText `json:"content"`
}

func (CalloutBlockPayload) blockPayload() {}
func (p CalloutBlockPayload) Validate() error {
	if (p.Kind != "INFO" && p.Kind != "NOTE" && p.Kind != "WARNING" && p.Kind != "TIP") || len(p.Title) > 240 {
		return errors.New("invalid callout")
	}
	return p.Content.Validate()
}

type TableBlockPayload struct {
	Caption string     `json:"caption,omitempty"`
	Headers []string   `json:"headers"`
	Rows    [][]string `json:"rows"`
}

func (TableBlockPayload) blockPayload() {}
func (p TableBlockPayload) Validate() error {
	if len(p.Caption) > 1000 || len(p.Headers) == 0 || len(p.Headers) > MaxTableColumns || len(p.Rows) == 0 || len(p.Rows) > MaxTableRows {
		return errors.New("invalid table")
	}
	for _, h := range p.Headers {
		if strings.TrimSpace(h) == "" || len(h) > 2000 {
			return errors.New("invalid table header")
		}
	}
	for _, r := range p.Rows {
		if len(r) != len(p.Headers) {
			return errors.New("table must be rectangular")
		}
		for _, c := range r {
			if len(c) > 2000 {
				return errors.New("table cell too large")
			}
		}
	}
	return nil
}

type DownloadBlockPayload struct {
	Asset       AssetReference `json:"asset"`
	Label       string         `json:"label"`
	Description string         `json:"description,omitempty"`
}

func (DownloadBlockPayload) blockPayload() {}
func (p DownloadBlockPayload) Validate() error {
	if e := p.Asset.Validate(); e != nil {
		return e
	}
	if !title(p.Label) || len(p.Description) > 4000 {
		return errors.New("invalid download")
	}
	return nil
}

type KnowledgeCheckBlockPayload struct {
	AssessmentKey string `json:"assessmentKey"`
}

func (KnowledgeCheckBlockPayload) blockPayload() {}
func (p KnowledgeCheckBlockPayload) Validate() error {
	k, e := NormalizeStructureKey(p.AssessmentKey)
	if e != nil || k != p.AssessmentKey {
		return errors.New("invalid knowledge check reference")
	}
	return nil
}

type DividerBlockPayload struct{}

func (DividerBlockPayload) blockPayload()   {}
func (DividerBlockPayload) Validate() error { return nil }
func title(s string) bool                   { return strings.TrimSpace(s) != "" && len(s) <= 240 }
func safeURL(s string) error {
	if len(s) == 0 || len(s) > MaxURLLength || strings.IndexFunc(s, func(r rune) bool { return r < 0x20 || r == 0x7f }) >= 0 {
		return errors.New("invalid content URL")
	}
	if strings.HasPrefix(s, "/") {
		if strings.HasPrefix(s, "//") {
			return errors.New("invalid content URL")
		}
		return nil
	}
	u, e := url.ParseRequestURI(s)
	if e != nil || u.Scheme != "https" || u.Host == "" {
		return errors.New("invalid content URL")
	}
	return nil
}

File: frontend/src/lesson/content.ts
import type { components } from '../api/generated'

const stableKeyPattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/
const maxBlocks = 200

export type RichText = { nodes: RichTextNode[] }
export type RichTextNode =
  | { type: 'paragraph'; content: RichTextInline[] }
  | { type: 'bullet_list' | 'ordered_list'; items: RichTextInline[][] }
export type RichTextInline = { type: 'text'; text: string; marks: RichTextMark[] } | { type: 'hard_break' }
export type RichTextMark = { type: 'emphasis' | 'strong' | 'inline_code' } | { type: 'link'; href: string }

type AssetReference = { assetKey: string }
export type RenderableBlock =
  | { key: string; type: 'TEXT'; payload: { content: RichText } }
  | { key: string; type: 'HEADING'; payload: { level: 2 | 3 | 4; content: RichTextInline[] } }
  | { key: string; type: 'IMAGE'; payload: { asset: AssetReference; decorative: boolean; altText?: string; caption?: string } }
  | { key: string; type: 'VIDEO'; payload: { asset: AssetReference; title: string; transcript?: string; transcriptAsset?: AssetReference; captionsAsset: AssetReference } }
  | { key: string; type: 'AUDIO'; payload: { asset: AssetReference; title: string; transcript?: string; transcriptAsset?: AssetReference } }
  | { key: string; type: 'CODE'; payload: { code: string; language?: string; title?: string } }
  | { key: string; type: 'QUOTE'; payload: { text: string; attribution?: string; sourceUrl?: string } }
  | { key: string; type: 'CALLOUT'; payload: { kind: 'INFO' | 'NOTE' | 'WARNING' | 'TIP'; title?: string; content: RichText } }
  | { key: string; type: 'TABLE'; payload: { caption?: string; headers: string[]; rows: string[][] } }
  | { key: string; type: 'DOWNLOAD'; payload: { asset: AssetReference; label: string; description?: string } }
  | { key: string; type: 'KNOWLEDGE_CHECK'; payload: { assessmentKey: string } }
  | { key: string; type: 'DIVIDER'; payload: Record<string, never> }
  | { key: string; type: 'UNSUPPORTED' }

export type DecodedLessonContent =
  | { kind: 'ready'; schemaVersion: 1; blocks: RenderableBlock[] }
  | { kind: 'unsupported-schema' }
  | { kind: 'invalid-content' }

// Generated types describe the HTTP contract. This narrow decoded view is the
// rendering boundary: it validates untrusted runtime JSON again, keeps Vue out
// of arbitrary payloads, and deliberately has no HTML/editor-state escape hatch.
export function decodeLessonContent(value: components['schemas']['LessonContent'] | unknown): DecodedLessonContent {
  if (!isRecord(value) || value.schemaVersion !== 1) return { kind: 'unsupported-schema' }
  if (!Array.isArray(value.blocks) || value.blocks.length > maxBlocks) return { kind: 'invalid-content' }

  const keys = new Set<string>()
  const blocks: RenderableBlock[] = []
  for (const rawBlock of value.blocks) {
    if (!isRecord(rawBlock) || !isStableKey(rawBlock.key) || keys.has(rawBlock.key)) return { kind: 'invalid-content' }
    keys.add(rawBlock.key)
    blocks.push(decodeBlock(rawBlock))
  }
  return { kind: 'ready', schemaVersion: 1, blocks }
}

export function safePublishedURL(value: string | undefined): string | undefined {
  if (!value || value.length > 2_048) return undefined
  if (value.startsWith('/')) {
    try {
      return new URL(value, 'https://academy.invalid').origin === 'https://academy.invalid' ? value : undefined
    } catch {
      return undefined
    }
  }
  try {
    return new URL(value).protocol === 'https:' ? value : undefined
  } catch {
    return undefined
  }
}

function decodeBlock(block: Record<string, unknown>): RenderableBlock {
  const key = block.key as string
  const payload = isRecord(block.payload) ? block.payload : undefined
  if (!payload || !isString(block.type)) return unsupported(key)

  switch (block.type) {
    case 'TEXT': {
      const content = richText(payload.content)
      return content ? { key, type: 'TEXT', payload: { content } } : unsupported(key)
    }
    case 'HEADING': {
      const level = payload.level
      const content = inlineList(payload.content)
      return (level === 2 || level === 3 || level === 4) && content ? { key, type: 'HEADING', payload: { level, content } } : unsupported(key)
    }
    case 'IMAGE': {
      const asset = assetReference(payload.asset)
      const decorative = payload.decorative
      const altText = optionalString(payload.altText)
      const caption = optionalString(payload.caption)
      if (!asset || typeof decorative !== 'boolean' || (decorative && altText) || (!decorative && !altText)) return unsupported(key)
      return { key, type: 'IMAGE', payload: { asset, decorative, ...(altText ? { altText } : {}), ...(caption ? { caption } : {}) } }
    }
    case 'VIDEO': {
      const asset = assetReference(payload.asset)
      const captionsAsset = assetReference(payload.captionsAsset)
      const title = optionalString(payload.title)
      const transcript = optionalString(payload.transcript)
      const transcriptAsset = assetReference(payload.transcriptAsset)
      if (!asset || !captionsAsset || !title || (!transcript && !transcriptAsset)) return unsupported(key)
      return { key, type: 'VIDEO', payload: { asset, title, captionsAsset, ...(transcript ? { transcript } : {}), ...(transcriptAsset ? { transcriptAsset } : {}) } }
    }
    case 'AUDIO': {
      const asset = assetReference(payload.asset)
      const title = optionalString(payload.title)
      const transcript = optionalString(payload.transcript)
      const transcriptAsset = assetReference(payload.transcriptAsset)
      if (!asset || !title || (!transcript && !transcriptAsset)) return unsupported(key)
      return { key, type: 'AUDIO', payload: { asset, title, ...(transcript ? { transcript } : {}), ...(transcriptAsset ? { transcriptAsset } : {}) } }
    }
    case 'CODE': {
      const code = optionalString(payload.code)
      const language = optionalString(payload.language)
      const title = optionalString(payload.title)
      return code !== undefined ? { key, type: 'CODE', payload: { code, ...(language ? { language } : {}), ...(title ? { title } : {}) } } : unsupported(key)
    }
    case 'QUOTE': {
      const text = optionalString(payload.text)
      const attribution = optionalString(payload.attribution)
      const sourceUrl = safePublishedURL(optionalString(payload.sourceUrl))
      return text ? { key, type: 'QUOTE', payload: { text, ...(attribution ? { attribution } : {}), ...(sourceUrl ? { sourceUrl } : {}) } } : unsupported(key)
    }
    case 'CALLOUT': {
      const kind = payload.kind
      const title = optionalString(payload.title)
      const content = richText(payload.content)
      if ((kind !== 'INFO' && kind !== 'NOTE' && kind !== 'WARNING' && kind !== 'TIP') || !content) return unsupported(key)
      return { key, type: 'CALLOUT', payload: { kind, ...(title ? { title } : {}), content } }
    }
    case 'TABLE': {
      const caption = optionalString(payload.caption)
      const headers = stringList(payload.headers)
      const rows = stringMatrix(payload.rows)
      if (!headers?.length || !rows || rows.some((row) => row.length !== headers.length)) return unsupported(key)
      return { key, type: 'TABLE', payload: { headers, rows, ...(caption ? { caption } : {}) } }
    }
    case 'DOWNLOAD': {
      const asset = assetReference(payload.asset)
      const label = optionalString(payload.label)
      const description = optionalString(payload.description)
      return asset && label ? { key, type: 'DOWNLOAD', payload: { asset, label, ...(description ? { description } : {}) } } : unsupported(key)
    }
    case 'KNOWLEDGE_CHECK': {
      const assessmentKey = optionalString(payload.assessmentKey)
      return assessmentKey ? { key, type: 'KNOWLEDGE_CHECK', payload: { assessmentKey } } : unsupported(key)
    }
    case 'DIVIDER': return { key, type: 'DIVIDER', payload: {} }
    default: return unsupported(key)
  }
}

function richText(value: unknown): RichText | undefined {
  if (!isRecord(value) || !Array.isArray(value.nodes)) return undefined
  const nodes: RichTextNode[] = []
  for (const rawNode of value.nodes) {
    if (!isRecord(rawNode) || !isString(rawNode.type)) return undefined
    if (rawNode.type === 'paragraph') {
      const content = inlineList(rawNode.content)
      if (!content) return undefined
      nodes.push({ type: 'paragraph', content })
    } else if (rawNode.type === 'bullet_list' || rawNode.type === 'ordered_list') {
      if (!Array.isArray(rawNode.items)) return undefined
      const items = rawNode.items.map(inlineList)
      if (items.some((item) => !item)) return undefined
      nodes.push({ type: rawNode.type, items: items as RichTextInline[][] })
    } else return undefined
  }
  return { nodes }
}

function inlineList(value: unknown): RichTextInline[] | undefined {
  if (!Array.isArray(value)) return undefined
  const result: RichTextInline[] = []
  for (const rawInline of value) {
    if (!isRecord(rawInline) || !isString(rawInline.type)) return undefined
    if (rawInline.type === 'hard_break') result.push({ type: 'hard_break' })
    else if (rawInline.type === 'text' && isString(rawInline.text)) {
      const marks = marksList(rawInline.marks)
      if (!marks) return undefined
      result.push({ type: 'text', text: rawInline.text, marks })
    } else return undefined
  }
  return result
}

function marksList(value: unknown): RichTextMark[] | undefined {
  if (value === undefined) return []
  if (!Array.isArray(value)) return undefined
  const result: RichTextMark[] = []
  for (const rawMark of value) {
    if (!isRecord(rawMark) || !isString(rawMark.type)) return undefined
    if (rawMark.type === 'emphasis' || rawMark.type === 'strong' || rawMark.type === 'inline_code') result.push({ type: rawMark.type })
    else if (rawMark.type === 'link') {
      const href = safePublishedURL(optionalString(rawMark.href))
      if (!href) return undefined
      result.push({ type: 'link', href })
    } else return undefined
  }
  return result
}

function assetReference(value: unknown): AssetReference | undefined {
  return isRecord(value) && isString(value.assetKey) && value.assetKey.length >= 1 && value.assetKey.length <= 160 ? { assetKey: value.assetKey } : undefined
}
function stringList(value: unknown): string[] | undefined { return Array.isArray(value) && value.every(isString) ? value : undefined }
function stringMatrix(value: unknown): string[][] | undefined { return Array.isArray(value) && value.every((row) => stringList(row)) ? value as string[][] : undefined }
function optionalString(value: unknown): string | undefined { return isString(value) ? value : undefined }
function isString(value: unknown): value is string { return typeof value === 'string' }
function isStableKey(value: unknown): value is string { return isString(value) && value.length >= 3 && value.length <= 96 && stableKeyPattern.test(value) }
function isRecord(value: unknown): value is Record<string, unknown> { return typeof value === 'object' && value !== null && !Array.isArray(value) }
function unsupported(key: string): RenderableBlock { return { key, type: 'UNSUPPORTED' } }
